pkg/render/render.go

```diff
@@ -250,3 +250,14 @@ func appendNotNil(components []Component, c Component) []Component {
 	}
 	return components
 }
+
+type K8sAttributes interface {
```

Comment:
Is this interface used? A quick search doesn't have it show up.

Looks like it has dex specific stuff in it but isn't in a dex specific package?

Reply:
Hey Casey, the interface is used in dex_config.go and implemented by 3 other types. The idea is to keep this generic and maybe make the constructors of render objects smaller across the project.

Reply:
The same items need to be rendered in a few different places, so we came up with this pattern so we could minimise the amount of duplicate around things like volume mounts, copying secrets, setting annotations, and figured this would be a great pattern to use in the future for other components that need these items rendered in multiple places.

Reply:
Gotcha - those files were minimized in my GH - I guess since the PR is so large :) 

I'm feeling like this needs to be scoped within a dex package, maybe along with all of the other dex utilities in dex_config.go?

Alternatively we could rename from K8sAttributes -> DexK8sAttributes or something like that, but it starts to be suspicious when a package needs to start prepending words to scope its structs / functions / etc. Usually a good sign that they deserve their own package.

Reply:
The K8sAttributes interface could be used in a variety of different ways, not just with dex. I think there are other components that could benefit from this. Elasticsearch is one that comes to mind, I previously used a decorator like pattern for this but it doesn't work as well as this.

Voltron could be another candidate, as I think similar set up needs to be done in multiple places, i.e. setting annotations, setting common env variables, setting volume mounts and volumes. Regardless, the point is that this really is a generic interface that anything component could take advantage of.

> it starts to be suspicious when a package needs to start prepending words to scope its structs / functions / etc. Usually a good sign that they deserve their own package

Frankly I think that ship has long since sailed with the render package, in my opinion that's quite a few packages mixed into one.

Reply:
This isn't a review of the entire render package, this is a review of a new interface whose comments specifically call out "dex" and whose name is _very_ generic. Just because the render packages is broad doesn't mean we should put every line of code into it. Let's acknowledge that it is too large and take steps to fix it!

This is just not a well defined interface at the moment, and it could easily be. Can we either:
- Remove the references to dex and re-brand this?
- More strongly associate this with dex built its own package, renaming it to be more specific?

Reply:
> The idea is to keep this generic and make the constructors of render objects smaller across the project.

I'm more than a little worried we're ending up with too many ways to do the same thing in the operator code. As it is, none of the other components take advantage of this and so rather than intorduce some new unifying simplicity, we've added additional complexity. 

if we hadn't already merged this, I would ask that either we revert this change and to it as a separate PR to _all of the components_ so that we achieve consistency, or that we achieve that consistency in this PR itself. 

Are you planning a follow-on PR to make sure this interface is used consistently?

Reply:
I think I should have been more clear with my comment on the package as a whole, and why the whole package would need to be tackled at once, in short it's because we would get cyclic imports. We likely want each component to be a sub package in render, however, there is common code that most every until uses that's in the render package itself. So if we move dex related stuff to it's own sub package it would refer to items in the base render package, but compliance, log storage, and manager would refer back to the dex package. I think this is a problem even if it's not sub packages. 

I had thought about this long ago with log storage and just came to this conclusion, so this isn't just a comment from me saying "oh it's already a problem everywhere else why fix it here", it's just seems like it is an all or nothing thing (at least for the easiest path)

As for the "well defined ness" of the interface, what's Dex specific about it? If you see something there then ya let's change it to be truly generic. 

I'm also wondering what you mean about "too many ways to do the same thing". I thought the only patterns we had was no pattern (pass every resource as it's own parameter) and the Elasticsearch decorator pattern. I'm all for following up and removing the decorator pattern and exchange it for this (I'm happy to take that on myself). I was more thinking there was no pattern like this in the render code so why not add this.

I'd like to admit as well that it was a real error in my judgement to not think about getting your input first before going through with this, and for that I am sorry, adding a new pattern should have been approved by you and or Erik.

Reply:
>  what's Dex specific about it?

```
	// RequiredSecrets returns secrets that you need to render for dex.
```

```
	// RequiredVolumes returns volumes that are related to dex.
```

```
	// RequiredEnv returns env that is used to configure pods with dex options.
```
 etc

Reply:
Right the comments, I didn't notice them. @rene-dekker could you update those please.

Was there anything dex specific about the interface itself?

Reply:
There was nothing dex specific about the lines of code that get compiled, no. But to be clear the comments are the interface - they define what the interface is and does. Without them it could pretty much do anything.

Reply:
> I'm also wondering what you mean about "too many ways to do the same thing".

This is probably an overly dramatic way to state what I was thinking. But essentially, if we're trying to introduce a new simplifying paradigm here, we need to roll it out everywhere. Otherwise the next person who comes along is just going to implement their own thing - either because they don't see this option because it's only used in one case (most likely) or because they're going to think "there's no established pattern so I will try to establish one that works for me"

I really appreciate the thought and effort going in to this, and don't have any problems with the actual implementation. But I really feel like we need to do this 100% or not at all, and right now I don't think this is completed.

Reply:
Ya I can definitely understand and agree with a) making sure it's visible to the next developer making changes and b) rolling it out completely. Not everything must use this pattern, but I think we should at least get rid of that previous Elasticsearch decorator pattern and replace it with this, it solved the problem of rendering the same secrets but not the problem of having to pass everything individually. 

So if you're OK with this I'm happy to take on the responsibility of removing that pattern as well as look into more specific package management after this release is through, should we agree with what we should do going forward. I'm also definitely OK with making the interface dex specific until then if that's what you want. WDYT?

Reply:
This is a great discussion. We have some real pain points in the code and we did an attempt to address it. For now I will rename the interface and move it to the dex_config.go file.

I am sure we now have a better perspective on this issue and an example of how to de-clutter that render function, so let's pick up this discussion elsewhere on a more suitable/interactive medium (video  call or dm) separate from this PR.

Reply:
That sounds good, thanks @rene-dekker.